pkg/executor/join/left_outer_semi_join_probe.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package join
+
+import (
+	"github.com/pingcap/tidb/pkg/expression"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/queue"
+	"github.com/pingcap/tidb/pkg/util/sqlkiller"
+)
+
+type leftOuterSemiJoinProbe struct {
+	baseJoinProbe
+
+	// isMatchedRows marks whether the left side row is matched
+	isMatchedRows []bool
+	// isNullRows marks whether the left side row matched result is null
+	isNullRows []bool
+
+	// buffer isNull for other condition evaluation
+	isNulls []bool
+
+	// used in other condition to record which rows need to be processed
+	processedProbeRowIdxQueue *queue.Queue[int]
+}
+
+var _ ProbeV2 = &leftOuterSemiJoinProbe{}
+
+func newLeftOuterSemiJoinProbe(base baseJoinProbe) *leftOuterSemiJoinProbe {
+	probe := &leftOuterSemiJoinProbe{
+		baseJoinProbe: base,
+	}
+	if base.ctx.hasOtherCondition() {
+		probe.processedProbeRowIdxQueue = queue.NewQueue[int](32)
+	}
+	return probe
+}
+
+func (j *leftOuterSemiJoinProbe) SetChunkForProbe(chunk *chunk.Chunk) (err error) {
+	err = j.baseJoinProbe.SetChunkForProbe(chunk)
+	if err != nil {
+		return err
+	}
+	j.isMatchedRows = j.isMatchedRows[:0]
+	for i := 0; i < j.chunkRows; i++ {
+		j.isMatchedRows = append(j.isMatchedRows, false)
+	}
+	j.isNullRows = j.isNullRows[:0]
+	for i := 0; i < j.chunkRows; i++ {
+		j.isNullRows = append(j.isNullRows, false)
+	}
+	if j.ctx.hasOtherCondition() {
+		j.processedProbeRowIdxQueue.Clear()
+		for i := 0; i < j.chunkRows; i++ {
```

Comment:
Since `matchedRowsHeaders` is already set in `base.SetChunkForProbe`, I think we can only add the rows that `matchedRowsHeaders[rows] != 0` to `processedProbeIdxQueue`.